services/system/Transact/plugin/src/helpers.rs

```diff
@@ -0,0 +1,171 @@
+use crate::bindings::accounts::plugin::accounts::{get_account, get_logged_in_user};
+use crate::bindings::accounts::smart_auth::smart_auth::{
+    self as SmartAuth, Action as PartialAction,
+};
+use crate::bindings::host::common as Host;
+use crate::errors::ErrorType::*;
+use crate::types::FromExpirationTime;
+use crate::AuthPlugins;
+use psibase::fracpack::Pack;
+use psibase::{
+    services, services::invite::action_structs as InviteService, AccountNumber, Hex, MethodNumber,
+    SignedTransaction, Tapos, Transaction,
+};
+
+use regex::Regex;
+use sha2::{Digest, Sha256};
+
+pub fn validate_action_name(action_name: &str) -> Result<(), Host::types::Error> {
+    let re = Regex::new(r"^[a-zA-Z0-9_]+$").unwrap();
+    if re.is_match(action_name) {
+        return Ok(());
+    }
+    Err(InvalidActionName.err(action_name))
+}
+
+pub fn assert_from_supervisor() {
+    let sender_app = Host::client::get_sender_app()
+        .app
+        .expect("Sender app not set");
+    assert!(sender_app == "supervisor", "Unauthorized");
+}
+
+fn inject_sender(actions: &[PartialAction], sender: AccountNumber) -> Vec<psibase::Action> {
+    actions
+        .into_iter()
+        .map(|a| psibase::Action {
+            sender,
+            service: AccountNumber::from(a.service.as_str()),
+            method: MethodNumber::from(a.method.as_str()),
+            rawData: Hex::from(a.raw_data.clone()),
+        })
+        .collect()
+}
+
+// The auth service can parameterize what claims to add to a transaction by both the sender
+//   and the actions in the transaction.
+// The auth service is therefore the only service other than transact that gets to see
+//   in full what actions are being taken by a user.
+fn get_claims(
+    sender: &AccountNumber,
+    actions: &[PartialAction],
+) -> Result<Vec<SmartAuth::Claim>, Host::types::Error> {
+    let sender_str = sender.to_string();
+    let mut claims: Vec<SmartAuth::Claim> = vec![];
+
+    // If the sender is not the invite payer, then we must get claims/proofs from their auth service
+    if *sender != *services::invite::PAYER_ACCOUNT {
```

Comment:
Yeah I like that, that is the intended behavior.

Note that for now I still haven't figured out how to actually eliminate hardcoding the `PAYER_ACCOUNT` in transact plugin entirely because of `get_tx_sender`.